File: leetcode/exercises/1.5.1.medium.top_k_frequent_elements_1.py
# -----------------------
# Created     : 21/10/2025
# Last Edited : 24/10/2025
# Topics      : 
# Big O       :
# Problem Id  : 347
# References  :
# Notes       :
# -----------------------

class Solution:
    def topKFrequent(self, nums: list[int], k: int) -> list[int]:

        map = {}
        for i in nums:
            map[i] = map.get(i, 0) + 1

        result = []
        map = sorted(map.items(), key=lambda x: x[1], reverse=True)
        for i in range(k):
            result.append(map[i][0])

        return result


sol = Solution()
print()
print(sol.topkfrequent([1,1,1,2,2,3], 2))

File: leetcode/exercises/1.5.1.medium.top_k_frequent_elements_1.go
// -----------------------
// Created     : 03/11/2025
// Last Edited : 03/11/2025
// Topics      : 
// Big O       :
// Problem Id  : 347
// References  :
// Notes       :
// -----------------------

package main

import (
    "fmt"
)

func topKFrequent(nums []int, k int) []int {

    counts := make(map[int]int)
    for _,v := range nums {
        if _, ok := counts[v]; !ok{
        counts[v] = -1
        }
        counts[v]++
    }

    freq := make([][]int, len(nums))
    for k,v := range counts {
        freq[v] = append(freq[v], k)
    }

    result := []int{}
    for i := len(freq) - 1; i >= 0; i-- {
        for _, v := range freq[i] {
            result = append(result, v)
            if len(result) == k { return result }
        }
    }

    return result
    
}

func main() {
    fmt.Println(topKFrequent([]int{1, 1, 3, 1, 3, 2, 10, 10, 9}, 3))
}
